// @nextcloud/dialogs/dist/chunks/DialogBase-XAlPDcV_.cjs
"use strict";
const c = require("vue"), u = require("@nextcloud/vue"), C = require("@vueuse/core"), h = /* @__PURE__ */ c.defineComponent({
  __name: "DialogButton",
  props: {
    callback: null,
    label: null,
    icon: null,
    type: null
  },
  emits: ["click"],
  setup(t, { emit: e }) {
    const o = t;
    return { __sfc: !0, props: o, emit: e, handleClick: (s) => {
      var l;
      (l = o.callback) == null || l.call(o), e("click", s);
    }, NcButton: u.NcButton, NcIconSvgWrapper: u.NcIconSvgWrapper };
  }
});
function f(t, e, o, n, s, l, _, p) {
  var a = typeof t == "function" ? t.options : t;
  e && (a.render = e, a.staticRenderFns = o, a._compiled = !0), n && (a.functional = !0), l && (a._scopeId = "data-v-" + l);
  var i;
  if (_ ? (i = function(r) {
    r = r || // cached call
    this.$vnode && this.$vnode.ssrContext || // stateful
    this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext, !r && typeof __VUE_SSR_CONTEXT__ < "u" && (r = __VUE_SSR_CONTEXT__), s && s.call(this, r), r && r._registeredComponents && r._registeredComponents.add(_);
  }, a._ssrRegister = i) : s && (i = p ? function() {
    s.call(
      this,
      (a.functional ? this.parent : this).$root.$options.shadowRoot
    );
  } : s), i)
    if (a.functional) {
      a._injectStyles = i;
      var d = a.render;
      a.render = function(m, g) {
        return i.call(g), d(m, g);
      };
    } else {
      var v = a.beforeCreate;
      a.beforeCreate = v ? [].concat(v, i) : [i];
    }
  return {
    exports: t,
    options: a
  };
}
var N = function() {
  var e = this, o = e._self._c, n = e._self._setupProxy;
  return o(n.NcButton, { attrs: { "aria-label": n.props.label, type: n.props.type }, on: { click: n.handleClick }, scopedSlots: e._u([n.props.icon !== void 0 ? { key: "icon", fn: function() {
    return [typeof n.props.icon == "string" ? o(n.NcIconSvgWrapper, { attrs: { svg: n.props.icon } }) : o(n.props.icon, { tag: "component", attrs: { size: 20 } })];
  }, proxy: !0 } : null], null, !0) }, [e._v(" " + e._s(n.props.label) + " ")]);
}, b = [], k = /* @__PURE__ */ f(
  h,
  N,
  b,
  !1,
  null,
  null,
  null,
  null
);
const y = k.exports, B = /* @__PURE__ */ c.defineComponent({
  __name: "DialogBase",
  props: {
    name: null,
    message: { default: "" },
    additionalTrapElements: { default: () => [] },
    container: { default: void 0 },
    size: { default: "small" },
    buttons: { default: () => [] },
    navigationClasses: { default: () => [] },
    contentClasses: { default: () => [] },
    dialogClasses: { default: () => [] }
  },
  emits: ["close"],
  setup(t, { emit: e }) {
    const o = t, n = c.useSlots(), s = c.ref(), { width: l } = C.useElementSize(s), _ = c.computed(() => l.value < 876), p = c.computed(() => (n == null ? void 0 : n.navigation) !== void 0), a = c.ref(!0), i = () => {
      a.value = !1, e("close");
    }, d = c.computed(() => ({
      additionalTrapElements: o.additionalTrapElements,
      container: o.container === void 0 ? "body" : o.container,
      name: o.name,
      size: o.size,
      enableSlideshow: !1,
      enableSwipe: !1
    }));
    return { __sfc: !0, props: o, emit: e, slots: n, wrapper: s, dialogWidth: l, isNavigationCollapsed: _, hasNavigation: p, open: a, handleClose: i, modalProps: d, NcModal: u.NcModal, DialogButton: y };
  }
});
var S = function() {
  var e = this, o = e._self._c, n = e._self._setupProxy;
  return n.open ? o(n.NcModal, e._b({ staticClass: "dialog__modal", on: { close: n.handleClose } }, "NcModal", n.modalProps, !1), [o("h2", { staticClass: "dialog__name", domProps: { textContent: e._s(e.name) } }), o("div", { staticClass: "dialog", class: e.dialogClasses }, [o("div", { ref: "wrapper", class: ["dialog__wrapper", { "dialog__wrapper--collapsed": n.isNavigationCollapsed }] }, [n.hasNavigation ? o("nav", { staticClass: "dialog__navigation", class: e.navigationClasses }, [e._t("navigation", null, { isCollapsed: n.isNavigationCollapsed })], 2) : e._e(), o("div", { staticClass: "dialog__content", class: e.contentClasses }, [e._t("default", function() {
    return [o("p", [e._v(e._s(n.props.message))])];
  })], 2)]), o("div", { staticClass: "dialog__actions" }, [e._t("actions", function() {
    return e._l(n.props.buttons, function(s, l) {
      return o(n.DialogButton, e._b({ key: l, on: { click: n.handleClose } }, "DialogButton", s, !1));
    });
  })], 2)])]) : e._e();
}, $ = [], w = /* @__PURE__ */ f(
  B,
  S,
  $,
  !1,
  null,
  "f925c3fb",
  null,
  null
);
const z = w.exports;
exports.DialogBase = z;
exports.normalizeComponent = f;
