// @nextcloud/dialogs/dist/chunks/DialogBase-aHPYSWdD.mjs
import { defineComponent as m, useSlots as h, ref as v, computed as d } from "vue";
import { NcButton as b, NcIconSvgWrapper as N, NcModal as k } from "@nextcloud/vue";
import { useElementSize as y } from "@vueuse/core";
const $ = /* @__PURE__ */ m({
  __name: "DialogButton",
  props: {
    callback: null,
    label: null,
    icon: null,
    type: null
  },
  emits: ["click"],
  setup(t, { emit: e }) {
    const a = t;
    return { __sfc: !0, props: a, emit: e, handleClick: (o) => {
      var l;
      (l = a.callback) == null || l.call(a), e("click", o);
    }, NcButton: b, NcIconSvgWrapper: N };
  }
});
function g(t, e, a, n, o, l, c, _) {
  var s = typeof t == "function" ? t.options : t;
  e && (s.render = e, s.staticRenderFns = a, s._compiled = !0), n && (s.functional = !0), l && (s._scopeId = "data-v-" + l);
  var i;
  if (c ? (i = function(r) {
    r = r || // cached call
    this.$vnode && this.$vnode.ssrContext || // stateful
    this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext, !r && typeof __VUE_SSR_CONTEXT__ < "u" && (r = __VUE_SSR_CONTEXT__), o && o.call(this, r), r && r._registeredComponents && r._registeredComponents.add(c);
  }, s._ssrRegister = i) : o && (i = _ ? function() {
    o.call(
      this,
      (s.functional ? this.parent : this).$root.$options.shadowRoot
    );
  } : o), i)
    if (s.functional) {
      s._injectStyles = i;
      var p = s.render;
      s.render = function(C, f) {
        return i.call(f), p(C, f);
      };
    } else {
      var u = s.beforeCreate;
      s.beforeCreate = u ? [].concat(u, i) : [i];
    }
  return {
    exports: t,
    options: s
  };
}
var S = function() {
  var e = this, a = e._self._c, n = e._self._setupProxy;
  return a(n.NcButton, { attrs: { "aria-label": n.props.label, type: n.props.type }, on: { click: n.handleClick }, scopedSlots: e._u([n.props.icon !== void 0 ? { key: "icon", fn: function() {
    return [typeof n.props.icon == "string" ? a(n.NcIconSvgWrapper, { attrs: { svg: n.props.icon } }) : a(n.props.icon, { tag: "component", attrs: { size: 20 } })];
  }, proxy: !0 } : null], null, !0) }, [e._v(" " + e._s(n.props.label) + " ")]);
}, w = [], B = /* @__PURE__ */ g(
  $,
  S,
  w,
  !1,
  null,
  null,
  null,
  null
);
const D = B.exports, R = /* @__PURE__ */ m({
  __name: "DialogBase",
  props: {
    name: null,
    message: { default: "" },
    additionalTrapElements: { default: () => [] },
    container: { default: void 0 },
    size: { default: "small" },
    buttons: { default: () => [] },
    navigationClasses: { default: () => [] },
    contentClasses: { default: () => [] },
    dialogClasses: { default: () => [] }
  },
  emits: ["close"],
  setup(t, { emit: e }) {
    const a = t, n = h(), o = v(), { width: l } = y(o), c = d(() => l.value < 876), _ = d(() => (n == null ? void 0 : n.navigation) !== void 0), s = v(!0), i = () => {
      s.value = !1, e("close");
    }, p = d(() => ({
      additionalTrapElements: a.additionalTrapElements,
      container: a.container === void 0 ? "body" : a.container,
      name: a.name,
      size: a.size,
      enableSlideshow: !1,
      enableSwipe: !1
    }));
    return { __sfc: !0, props: a, emit: e, slots: n, wrapper: o, dialogWidth: l, isNavigationCollapsed: c, hasNavigation: _, open: s, handleClose: i, modalProps: p, NcModal: k, DialogButton: D };
  }
});
var T = function() {
  var e = this, a = e._self._c, n = e._self._setupProxy;
  return n.open ? a(n.NcModal, e._b({ staticClass: "dialog__modal", on: { close: n.handleClose } }, "NcModal", n.modalProps, !1), [a("h2", { staticClass: "dialog__name", domProps: { textContent: e._s(e.name) } }), a("div", { staticClass: "dialog", class: e.dialogClasses }, [a("div", { ref: "wrapper", class: ["dialog__wrapper", { "dialog__wrapper--collapsed": n.isNavigationCollapsed }] }, [n.hasNavigation ? a("nav", { staticClass: "dialog__navigation", class: e.navigationClasses }, [e._t("navigation", null, { isCollapsed: n.isNavigationCollapsed })], 2) : e._e(), a("div", { staticClass: "dialog__content", class: e.contentClasses }, [e._t("default", function() {
    return [a("p", [e._v(e._s(n.props.message))])];
  })], 2)]), a("div", { staticClass: "dialog__actions" }, [e._t("actions", function() {
    return e._l(n.props.buttons, function(o, l) {
      return a(n.DialogButton, e._b({ key: l, on: { click: n.handleClose } }, "DialogButton", o, !1));
    });
  })], 2)])]) : e._e();
}, z = [], E = /* @__PURE__ */ g(
  R,
  T,
  z,
  !1,
  null,
  "f925c3fb",
  null,
  null
);
const M = E.exports;
export {
  M as D,
  g as n
};
